File: frontend/src/api/sync.ts
import { apiRequest } from "./client";
import { LocalTask, SyncBootstrapResponse, SyncTaskRecord } from "../types/task";

export interface SyncTaskPayload {
  title: string;
  description?: string | null;
  type: LocalTask["type"];
  status: LocalTask["status"];
  deadline_at: string | null;
  remind_at: string | null;
  reminder_mode: LocalTask["reminder_mode"];
  reminder_time_local: string | null;
  reminder_interval_hours: number | null;
  recurrence_rule: string | null;
  // Client-captured snapshot of the user's reminder-shaping settings, so the
  // backend can compute/fire this task's reminders without storing any settings.
  reminder_timezone: string;
  reminder_language: string;
  snooze_minutes: number;
  updated_at: string;
  deleted_at: string | null;
}

interface SyncTaskUpsertResponse {
  applied: boolean;
  task: SyncTaskRecord;
}

interface SyncBatchPayload {
  tasks: Array<SyncTaskPayload & { client_task_id: string }>;
}

export function bootstrapSync(): Promise<SyncBootstrapResponse> {
  return apiRequest<SyncBootstrapResponse>("/sync/bootstrap");
}

export function listSyncChanges(since: string): Promise<SyncBootstrapResponse> {
  const params = new URLSearchParams({ since });
  return apiRequest<SyncBootstrapResponse>(`/sync/changes?${params.toString()}`);
}

export function syncTask(clientTaskId: string, payload: SyncTaskPayload): Promise<SyncTaskUpsertResponse> {
  return apiRequest<SyncTaskUpsertResponse>(`/sync/tasks/${clientTaskId}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function deleteSyncTask(clientTaskId: string): Promise<SyncTaskUpsertResponse> {
  return apiRequest<SyncTaskUpsertResponse>(`/sync/tasks/${clientTaskId}`, {
    method: "DELETE",
  });
}

export function syncTaskBatch(payload: SyncBatchPayload): Promise<SyncBootstrapResponse> {
  return apiRequest<SyncBootstrapResponse>("/sync/batch", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

File: backend/app/schemas/sync.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.task import ReminderMode, TaskStatus, TaskType


class SyncTaskBase(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=5000)
    type: TaskType
    status: TaskStatus = TaskStatus.active
    deadline_at: datetime | None = None
    remind_at: datetime | None = None
    reminder_mode: ReminderMode = ReminderMode.none
    reminder_time_local: str | None = Field(default=None, pattern=r"^(?:[01]\d|2[0-3]):[0-5]\d$")
    reminder_interval_hours: int | None = Field(default=None, ge=1, le=24)
    recurrence_rule: str | None = Field(default=None, max_length=255)
    # Client-captured snapshot of the user's reminder-shaping settings. The backend
    # uses these to compute/fire this task's reminders; it keeps no user settings.
    reminder_timezone: str | None = Field(default=None, max_length=64)
    reminder_language: str | None = Field(default=None, max_length=8)
    snooze_minutes: int | None = Field(default=None, ge=5, le=240)
    updated_at: datetime
    deleted_at: datetime | None = None


class SyncTaskUpsert(SyncTaskBase):
    pass


class SyncBatchItem(SyncTaskBase):
    client_task_id: str = Field(min_length=1, max_length=64)


class SyncTaskRecord(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    client_task_id: str
    title: str
    description: str | None
    type: TaskType
    status: TaskStatus
    deadline_at: datetime | None
    remind_at: datetime | None
    reminder_mode: ReminderMode
    reminder_time_local: str | None
    reminder_interval_hours: int | None
    recurrence_rule: str | None
    updated_at: datetime
    deleted_at: datetime | None
    completed_at: datetime | None
    cancelled_at: datetime | None
    last_reminded_at: datetime | None


class SyncTaskUpsertResponse(BaseModel):
    applied: bool
    task: SyncTaskRecord


class SyncBatchRequest(BaseModel):
    tasks: list[SyncBatchItem]


class SyncTaskListResponse(BaseModel):
    items: list[SyncTaskRecord]
    server_time: datetime
